File: Desktop/New folder/QuyHoachDong/BacThang.java
package QuyHoachDong;

import java.util.Scanner;

public class BacThang {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        while (t--> 0) {
            int n = scanner.nextInt();
            int k = scanner.nextInt();
            System.out.println(qhd(n, k));
        }
    }

    private static int qhd(int n, int k) {
        int i, j, f[] = new int[n+1], mod = (int) 1e9+7;
        f[0] = 1; f[1] = 1;
        for ( i = 2; i <= n; i++) {
            for (j= 1; j <= Math.min(k, i); j++) {
                f[i] = (f[i] +f[i-j]) %mod;
            }
        }
        return f[n];
    }
}

File: cpp/BacThang.cpp
#include<bits/stdc++.h>
using namespace std;
int M = 1e9+7;

int qhd(int n, int k){
    int F[n+1] = {0}, i, j;
    F[0]=1; F[1]=1;
    for(i = 2; i<=n; i++){
        for(j=1;j<=min(i,k);j++){
            F[i] = (F[i] + F[i-j])%M;
        }
    }
    return F[n];
}

main(){
    int t,n,k;
    cin>>t;
    while(t--){
        cin>>n>>k;
        cout<<qhd(n,k)<<endl;
    }
}
